### DISCUSSION TITLE
PostgrestException: ReferenceAttribute must be used with Postgrest BaseModels (supabase-csharp)

### DISCUSSION THREAD

**Discussion by will11600** [supabase-community/supabase-csharp#76] [Q&A] [answered]
Hi All,

I'm currently building a C# WebAssembly application using the [supabase-csharp library](https://github.com/supabase-community/supabase-csharp/tree/master). I have been getting this error every time I try to get data from a table. I have been debugging for a while now and I'm still scratching my head, was wondering if anyone might be able to point me in the right direction. I'm not sure what I'm doing wrong.

```
System.Reflection.CustomAttribute.GetCustomAttributesBase(ICustomAttributeProvider obj, Type attributeType, Boolean inheritedOnly) at System.Reflection.CustomAttribute.GetCustomAttributes(ICustomAttributeProvider obj, Type attributeType, Boolean inherit) at System.Reflection.RuntimePropertyInfo.GetCustomAttributes(Type attributeType, Boolean inherit) at Postgrest.Table`1[[SimonApp.Data.Report, SimonApp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]..ctor(String baseUrl, JsonSerializerSettings serializerSettings, ClientOptions options) at Postgrest.Client.Table[Report]() at SimonApp.Pages.Reports.GetReportsAsync() in C:\Users\mrnic\source\repos\SimonApp\Client\Pages\Reports.razor:line 46
```

The method that is throwing the error:
```cs
private async Task<Report[]> GetReportsAsync()
{
    IsLoading = true;

    try
    {
        var reports = await supabase.Postgrest.Table<Report>().Get();
        return reports.Models.ToArray();
    }
    catch (Exception ex) { exceptionHandler.Handle(ex); }
    finally { IsLoading = false; }

    return Array.Empty<Report>();
}
```

This is the class I am trying to retrieve, I have followed [these examples](https://github.com/supabase-community/supabase-csharp/tree/master/Examples/SupabaseExample/Models). However, mine does not seem to work.

```cs
using Postgrest.Attributes;
using Postgrest.Models;
using Supabase.Gotrue;

namespace SimonApp.Data
{
    [Table("reports")]
    public class Report : BaseModel
    {
        [PrimaryKey("id")]
        public int Id { get; set; }

        [Column("created_at")]
        public DateTime CreatedAt { get; set; }

        [Reference(typeof(Room), propertyName: "room_id")]
        public Room Room { get; set; } = default!;

        [Column("description")]
        public string Description { get; set; } = default!;

        [Reference(typeof(ServiceCategory), propertyName: "service_category_id")]
        public ServiceCategory ServiceCategory { get; set; } = default!;

        [Column("resolve_by_date")]
        public DateTime ResolveBy { get; set; }

        [Reference(typeof(User), propertyName: "user_uuid")]
        public User? User { get; set; }

        [Reference(typeof(ReportSeries), propertyName: "series_id")]
        public ReportSeries Series { get; set; } = default!;

        [Reference(typeof(ReportEvent))]
        public List<ReportEvent> Events { get; set; } = default!;
    }
}
```

Any help would be greatly appreciated! :)

**Comment by acupofjose** [comment]
Does the `ServiceCategory` class derive from `BaseModel`?

**Reply by will11600** [reply]
Thank you for getting in touch!

Yes, I can confirm all of the classes, including `ServiceCategory` derive from `BaseModel`.

**Reply by acupofjose** [reply]
Hm... a few other things to check:
- `ServiceCategory` (or another `Reference`-ed) class may also contain `Reference` attributes and one of those models may not be derived from `BaseModel`
- `Reference` could be referencing an attribute from another library, be sure that it's actually using `Postgrest.Attributes` (it looks like you are, but might as well confirm).

**Reply by will11600** [reply, marked as answer]
Ok, so I found the source of the issue! I had created a reference to `Supabase.Gotrue.User`, which does not derive from the `BaseModel` class.

What would be the best practice method of storing a reference to a user inside a table?

**Reply by acupofjose** [reply]
Ah! Interesting, yes. Store a foreign key on your desired table that references the Auth.User table. (i.e. Report has a `user_id` field that has a foreign key relationship to `Auth.User.Id`. Then create a reference model for User that derives from `BaseModel` with the columns you want selected.

**Reply by acupofjose** [reply]
Actually, I’m wrong there. You’d need a trigger that would move “public” data from Auth.User data into a Public.User table. I don’t believe postgrest allows you to query the Auth.User table from the client!

**Reply by will11600** [reply]
Ah ok, fair enough! Thanks for your help! :)

I thought about it, and actually, I don't think it's critical that I have access to that data from my client anyway.
The main purpose was so that I could display who had submitted each report in my application. While I was setting everything up, it didn't actually occur to me that, understandably so, there aren't columns for first name, last name, etc. in the `user` table out of the box.

On the online portal, the `auth` schema is listed under the protected category. So, I'm assuming that I'm not supposed to touch it? My plan is to set up an RLS-protected table under the `public` schema for this information. Unless there's a better way to do it?

**Reply by acupofjose** [reply]
Happy to! Yeah the suggested way would be to have a trigger for user registration that writes public data over to a different table. Say, for instance, a `profile` table. 

You can see more here! https://supabase.com/docs/guides/auth/managing-user-data

**Comment by will11600** [comment]
Thanks again!

I haven't fully implemented a user sign-up yet. However, I've taken your suggestion and setup a `profile` table!
